docs/string-representation.md

```diff
@@ -0,0 +1,549 @@
+    // Iterate over `src` taking into account ill-formed sequences.
+
+    // Inside the loop, call by `char` into property lookup shared by all of these.
+    // In case of capital sigma, scan forward in an inner loop.
+
+    // Including placeholder code that compiles.
+    (src.len(), dst.len())
+}
+
+/// Reads Latin1 from `src` and writes corresponding
+/// lower-cased text into `dst` as Latin1. Returns the
+/// number of code units read and the number of code units written. The
+/// number of code units read is guaranteed to be `src.len()` if
+/// `dst.len() >= to_lowercase_latin1_max(src.len()).unwrap()`.
+pub fn to_lowercase_latin1(src: &[u8], dst: &mut [u8]) -> (usize, usize) {
+    // Iterate over `src`.
+
+    // Inside the loop, index into a by-byte lookup table.
+
+    // Including placeholder code that compiles.
+    (src.len(), dst.len())
+}
+
+/// Some operation whose output size doesn't have a reasonable
+/// estimator function.
+pub fn inestimable(src: &str) -> String {
+    // Placeholder implementation that compiles
+    String::with_capacity(src.len())
+}
+
+/// Some operation whose output size doesn't have a reasonable
+/// estimator function. (Version that accepts potentially-invalid
+/// UTF-8.)
+///
+/// The returned `String` may have excess capacity. If you are holding onto
+/// it for an extended period of time, consider calling `shrink_to_fit()`
+/// on it.
+pub fn inestimable_utf8(src: &[u8]) -> String {
+    // Placeholder implementation that compiles
+    String::with_capacity(src.len())
+}
+
+/// Some operation whose output size doesn't have a reasonable
+/// estimator function. (Version that accepts potentially-invalid
+/// UTF-16.)
+///
+/// The returned `Vec` may have excess capacity. If you are holding onto
+/// it for an extended period of time, consider calling `shrink_to_fit()`
+/// on it.
+pub fn inestimable_utf16(src: &[u16]) -> Vec<u16> {
+    // Placeholder implementation that compiles
+    Vec::with_capacity(src.len())
+}
+
+// End functions that actually require thinking
+
+// Start Rust-only boilerplate
+
+/// Returns a `String` whose contents are `src` lower-cased.
+///
+/// The returned `String` may have excess capacity. If you are holding onto
+/// it for an extended period of time, consider calling `shrink_to_fit()`
+/// on it.
+pub fn to_lowercase(src: &str) -> String {
+    // This is the same boilerplate for all String`-returning wrappers
+    // and should be generated using a macro.
+    let mut vec = Vec::new();
+    // First assume the same length
+    vec.resize(src.len(), 0);
+    let (read_first, written_first) = to_lowercase_str_utf8(src, &mut vec);
+    let mut total_written = written_first;
+    let input_left = src.len() - read_first;
+    if input_left != 0 {
+        vec.resize(
+            written_first + to_lowercase_utf8_max(input_left).unwrap(),
+            0,
+        );
+        let (read_second, written_second) = to_lowercase_str_utf8(src, &mut vec);
```

Comment:
This should be `to_lowercase_str_utf8(&src[read_first..], &mut vec[written_first..])` right?

Reply:
Whoa. Yes. Thanks.